### ISSUE TITLE
Support large number of streams in blob storage destinations

### ISSUE THREAD

**Issue by tuliren** [airbytehq/airbyte#9586]
## Tell us about the problem you're trying to solve
- Currently we use one `StreamTransferManager` per stream to upload data to S3 and GCS.
- By default, each `StreamTransferManager` requires [31 MB](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-s3/src/main/java/io/airbyte/integrations/destination/s3/util/S3StreamTransferManagerHelper.java#L19) of memory to do the work.
- This is problematic when there are a large number of streams. The destination container will run into out-of-memory error (OOME).
- It is unclear how much memory a container can get. According to [Datadog](https://app.datadoghq.com/orchestration/overview/pod?query=destination-bigquery-sync-34986-0-kixhr&highlight=W1sic3BlYyIsImNvbnRhaW5lcnMiLCI0IiwicmVzb3VyY2VzIiwicmVxdWVzdHMiLCJtZW1vcnkiXV0%3D&inspect=04586ad2-6675-473e-8af2-6dd128b9bfcc&panel_tab=metrics&start=1642546247839&end=1642547147839&paused=false), one of the GCS container got OOME with 1.81 GB memory.
<img width="650" alt="Screen Shot 2022-01-18 at 15 03 43" src="https://user-images.githubusercontent.com/1933157/150034339-87a50955-4720-480f-828d-db80461decbd.png">

- This issue is related to https://github.com/airbytehq/oncall/issues/109.
- It affects both blob storage destinations, as well warehouse destinations that use a staging bucket.

## Describe the solution you’d like
- Unclear

## Describe the alternative you’ve considered or used
N/A

## Additional context
N/A

**Comment by edgao** [comment]
could this be as simple as just closing the previous StreamTransferManager whenever we open a new one? (e.g. [here](https://github.com/airbytehq/airbyte/blob/1832c69c269e1c9f15c3eb36f7a1bd59a861ab73/airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/s3/S3StreamCopier.java#L98-L113))

It looks like we only keep the writer open in case we want to write to a previous file, but that's never intentionally happening AFAICT. I think the only potential weird thing is if [this](https://github.com/airbytehq/airbyte/blob/dfb7b8cd3f22426ed0bc55f7b7fe83de4524a1d6/airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/copy/CopyConsumerFactory.java#L90) is happening multithreadedly, in which case the `fileName` tracking probably should be pushed down into `StreamCopier#write`.

**Comment by tuliren** [comment]
> could this be as simple as just closing the previous StreamTransferManager whenever we open a new one? (e.g. here)

Yes, totally possible. We just need to make sure we are always emitting streams sequentially on the source side. However, theoretically, it is possible to have multiple streams emitting messages are the same time.
